tinynn/graph/quantization/quantizable/gru.py

```diff
@@ -0,0 +1,341 @@
+            observed.layers[idx] = _GRULayer.from_float(other, idx, qconfig, batch_first=False)
+        observed.train()
+        observed = torch.ao.quantization.prepare_qat(observed, inplace=True)
+        return observed
+
+
+class GRUCell(torch.nn.Module):
+    r"""A quantizable long short-term memory (LSTM) cell.
+    For the description and the argument types, please, refer to :class:`~torch.nn.GRUCell`
+    """
+    _FLOAT_MODULE = torch.nn.GRUCell
+
+    def __init__(self, input_dim: int, hidden_dim: int, bias: bool = True,
+                 device=None, dtype=None) -> None:
+        factory_kwargs = {'device': device, 'dtype': dtype}
+        super().__init__()
+        self.input_size = input_dim
+        self.hidden_size = hidden_dim
+        self.bias = bias
+
+        self.igates = torch.nn.Linear(input_dim, 3 * hidden_dim, bias=bias, **factory_kwargs)
+        self.hgates = torch.nn.Linear(hidden_dim, 3 * hidden_dim, bias=bias, **factory_kwargs)
+
+        self.add1 = torch.ao.nn.quantized.FloatFunctional()
+        self.add2 = torch.ao.nn.quantized.FloatFunctional()
+        self.add3 = torch.ao.nn.quantized.FloatFunctional()
+        self.add4 = torch.ao.nn.quantized.FloatFunctional()
+        self.sub1 = torch.ao.nn.quantized.FloatFunctional()
+
+        self.mul1 = torch.ao.nn.quantized.FloatFunctional()
+        self.mul2 = torch.ao.nn.quantized.FloatFunctional()
+        self.mul3 = torch.ao.nn.quantized.FloatFunctional()
+
+        self.act1 = torch.nn.Sigmoid()
+        self.act2 = torch.nn.Tanh()
+        self.hidden_state_dtype: torch.dtype = torch.quint8
+
+    def forward(self, x: Tensor, hidden: Optional[Tensor] = None) -> Optional[Tuple[Tensor, Tensor], Tensor]:
+
+        result = []
+        if hidden is None or hidden[0] is None:
+            hidden = self.initialize_hidden(x.shape[0], x.is_quantized)
+
+        ri, zi, ni = self.igates(x).chunk(3, -1)
+        if x.dim() > 2:
+            for k in range(x.size(0)):
+                hx = hidden
+                rh, zh, nh = self.hgates(hx).chunk(3, -1)
+                rgate = self.act1(self.add1.add(ri[k, ...], rh))
+                zgate = self.act1(self.add2.add(zi[k, ...], zh))
+                ngate = self.act2(self.add3.add(ni[k, ...], self.mul1.mul(rgate, nh)))
+                hidden = self.add4.add(self.mul2.mul(self.sub1.add_scalar(torch.tensor(1), -zgate), ngate),
+                                       self.mul3.mul(zgate, hx))
+                result.append(hidden)
+            result_tensor = torch.stack(result, 0)
+            return result_tensor, hidden
+        else:
+            hx = hidden
+            rh, zh, nh = self.hgates(hx).chunk(3, -1)
+            rgate = self.act1(self.add1.add(ri, rh))
+            zgate = self.act1(self.add2.add(zi, zh))
+            ngate = self.act2(self.add3.add(ni, self.mul1.mul(rgate, nh)))
+            hidden = self.add4.add(self.mul2.mul(self.sub1.add_scalar(torch.tensor(1), -zgate), ngate),
```

Comment:
The usage is incorrect. It should be `self.sub1.add_scalar(-zgate, 1)`. BTW, I'm not sure if the negative operator`-zgate` is quantizable in legacy PyTorch. Maybe using another FloatFunctional to perform `.mul_scalar(zgate, -1)` is safer.